CMake/Findmlpack.cmake

```diff
@@ -0,0 +1,220 @@
+
+mark_as_advanced(ENSMALLEN_INCLUDE_DIR)
+
+if (ENSMALLEN_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${ENSMALLEN_INCLUDE_DIR}")
+else()
+  message(FATAL_ERROR 
+    "Ensmallen library is not found, please make sure to install ensmallen,"
+    " or use the autodownloader. mlpack can not be installed without cereal")
+endif()
+
+if (USE_OPENMP)
+  find_package(OpenMP)
+endif ()
+
+if (OpenMP_FOUND AND OpenMP_CXX_VERSION VERSION_GREATER_EQUAL 3.0.0)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+  set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${OpenMP_CXX_LIBRARIES})
+else ()
+  # Disable warnings for all the unknown OpenMP pragmas.
+  if (NOT MSVC)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas")
+  else ()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd4068")
+  endif ()
+  set(OpenMP_CXX_FLAGS "")
+endif ()
+
+# This module sets the following variable:
+#  MLPACK_FOUND - set to true if the library is found
+#  MLPACK_INCLUDE_DIR - list of required include directories
+#  MLPACK_VERSION_MAJOR - major version number
+#  MLPACK_VERSION_MINOR - minor version number
+#  MLPACK_VERSION_PATCH - patch version number
+#  MLPACK_VERSION_STRING - version number as a string (ex: "1.0.4")
+
+# This will be executed if mlpack is installed already.
+find_path(MLPACK_INCLUDE_DIR
+  NAMES mlpack.hpp
+  PATHS ${MLPACK_SEARCH_PATHS}/include)
+
+# This will be executed when compiling mlpack bindings and tests.
+if (NOT MLPACK_INCLUDE_DIR)
+  find_path(MLPACK_INCLUDE_DIR
+    NAMES mlpack.hpp
+    PATHS ${CMAKE_CURRENT_SOURCE_DIR}/src/)
+endif()
+
+if (MLPACK_INCLUDE_DIR)
+  # ------------------------------------------------------------------------
+  #  Extract version information from <mlpack>
+  # ------------------------------------------------------------------------
+
+  set(MLPACK_VERSION_MAJOR 0)
+  set(MLPACK_VERSION_MINOR 0)
+  set(MLPACK_VERSION_PATCH 0)
+
+  if (EXISTS "${MLPACK_INCLUDE_DIR}/mlpack/core/util/version.hpp")
+
+    set(MLPACK_FOUND YES)
+
+    # Read and parse armdillo version header file for version number
+    file(READ "${MLPACK_INCLUDE_DIR}/mlpack/core/util/version.hpp"
+        _mlpack_HEADER_CONTENTS)
+    string(REGEX REPLACE ".*#define MLPACK_VERSION_MAJOR ([0-9]+).*" "\\1"
+        MLPACK_VERSION_MAJOR "${_mlpack_HEADER_CONTENTS}")
+    string(REGEX REPLACE ".*#define MLPACK_VERSION_MINOR ([0-9]+).*" "\\1"
+        MLPACK_VERSION_MINOR "${_mlpack_HEADER_CONTENTS}")
+    string(REGEX REPLACE ".*#define MLPACK_VERSION_PATCH ([0-9]+).*" "\\1"
+        MLPACK_VERSION_PATCH "${_mlpack_HEADER_CONTENTS}")
+
+  endif()
+
+  set(MLPACK_VERSION_STRING "${MLPACK_VERSION_MAJOR}.${MLPACK_VERSION_MINOR}.${MLPACK_VERSION_PATCH}")
+endif ()
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(mlpack
+  REQUIRED_VARS MLPACK_INCLUDE_DIR
+  VERSION_VAR MLPACK_VERSION_STRING)
+
+mark_as_advanced(MLPACK_INCLUDE_DIR)
```

Comment:
I think we should mark `MLPACK_INCLUDE_DIRS` as advanced too, as well as `MLPACK_LIBRARIES`.